cilium/CFP-22215-mutual-auth-for-service-mesh.md

```diff
@@ -0,0 +1,442 @@
+  * Thinking of IP recycling case, e.g. in ENI mode an IP could move from one node to another.
+  * Trigger on CiliumEndpoint deletion even for pods on remote nodes, delete authentication state corresponding to that peer. Including userspace authentication table entries, datapath CT/map entries.
+
+##### Timer-based
+* Do we need periodic garbage collection, e.g. of authenticated sessions? Because the auth table has an expiry time, we have built-in garbage collection - the userspace will be responsible for pruning expired auth table entries if the identity has been deleted and the certificate validity period has passed.
+* At this stage, it seems likely that deletion of expired entries will still be done by userspace, although this could change in the future.
+
+#### Certificate management
+
+Initial certificate management for this feature will be using SPIRE and SPIFFE, with interfaces built to allow us to walk back to other options once we have more user intent data.
+
+### Data Plane
+
+
+#### Structures
+Policy Entry - 1 bit: "Authentication Required"
+Drop events - 1 codepoint - "Authentication Required"
+Dedicated bpf auth map, see Key question: How to implement dataplane auth check
+
+#### Logic
+When performing policy lookup, check 'authentication required' bit.
+If this bit is set and there is no corresponding entry in the BPF auth table, drop the traffic and emit an 'authentication required' event to the local control plane.
+Ensure both egress and ingress implement this logic.
+
+#### Proxy
+For an initial target, do not support Envoy participating in the service mesh, although the amount of work can be checked once we've done other things.
+
+### Lifecycle / operations walkthrough
+
+#### Connection
+From scratch, the following example assumes that SPIRE has been bootstrapped in the cluster and the spire-agent running as a container within the cilium-agent pod on each node. Additionally, a CiliumNetworkPolicy must be inserted that selects the target pods and enables the authentication options. We assume that these changes have been calculated and applied to the policymap entries for the pod in the example below.
+
+When a pod sends a packet for a new connection that is selected by a CiliumNetworkPolicy which marks that connectivity as requiring authentication, the BPF datapath goes through a series of steps, from evaluating the network policy to checking the Authentication map (auth). When the packet hits the policymap lookup, the map entry will signal that the connection requires authentication. This triggers an additional check in the datapath into the bpf auth map to determine the authentication status between the peers involved in the connection. For the example below, we assume that this initial authentication check fails because the authentication has not yet occurred for these peers.
+
+![SPIFFE and SPIRE Architecture with Cilium Datapath](./images/SPIFFE-SPIRE-arch-with-datapath.png)
```

Comment:
Small detail: the mTLS box will talk to the SPIFFE box as it is the one that needs (and knows how to work with) the certificates.

Reply:
Thanks, updated!